#12 implement email confirmation

DMCHMB opened the pull request.

IliaWizy commented on .github/workflows/workflowForDevelop.yml:
```diff
@@ -3,7 +3,7 @@ name: Develop branch workflow
   push:
     branches:
-      - dev
+      - feature/*
```
лишний пуш на промежуточных, не окончательных действия - это ресурсы. ресурсы стоят денег. иногда одна строка может стоить миллионы. пуш на ветку фичи, чтобы собирать артифакт? где его потом используют, что такая потребность?

IliaWizy commented on config/checkstyle/checkstyle.xml:
```diff
@@ -62,7 +62,7 @@
             <property name="allowNonPrintableEscapes" value="true"/>
         </module>
-        <module name="AvoidStarImport"/>
+<!--        <module name="AvoidStarImport"/>-->
```
https://checkstyle.sourceforge.io/checks/imports/avoidstarimport.html или донастроить для библиоткет или оставить. но для использования импорта всего пакета - нет.

sursindmitry replied:
А не проще оставить Гугловский чекстайл и сделать, что бы IDE не делала эти звёздочки?
![image](https://github.com/IliaWizy/education-app/assets/40328886/1cf1b89b-d327-47a2-b26c-8116b78be9be)

IliaWizy replied:
да. отличное решение, чтобы иде не объединяло

DMCHMB replied:
Спасибо! Я даже занесу себе в заметки

IliaWizy commented on http/register_user_duplicate.http:
как пишешь unit тест? один класс - один класс теста. тут почем N файлов для теста?

DMCHMB replied:
не разобрался, как написать тест по активации пользователя (как получить емаил-токен)

IliaWizy commented on src/main/java/com/wizy/educationapp/AppStarter.java:
```diff
@@ -2,9 +2,8 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
-import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
 
-@SpringBootApplication(exclude = {DataSourceAutoConfiguration.class})
```
миграцию также можно было отключить. 
https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/autoconfigure/liquibase/LiquibaseAutoConfiguration.html

DMCHMB replied:
Как делаем? Отключаем миграцию, либо делаем как в dev:
exclude = {DataSourceAutoConfiguration.class
?

IliaWizy replied:
на откуп тебе. в итоге, как эксперимент, если останется время - попробуй придумать как реализовать миграцию для прода и разработки одновременно. если нет - потом расскажу про этот чит

IliaWizy commented on src/main/java/com/wizy/educationapp/controller/AuthorizationController.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/api/v1")
```
- /api излишне. обычно они на домене/ поддомене
- AuthorizationController что говорит его имя? контроллер авторизация, а значит - /v1/... - тк класс название класса должно давать понять что внутри него и группировать методы контроллера

IliaWizy commented on src/main/java/com/wizy/educationapp/controller/AuthorizationController.java:
```diff
@@ -0,0 +1,44 @@
+  private final UserMapper userMapper;
+  private final VerificationTokenService verificationTokenService;
+
+  @PostMapping("/auth/sign-up")
```
см выше. оставляем?

IliaWizy commented on src/main/java/com/wizy/educationapp/controller/AuthorizationController.java:
```diff
@@ -0,0 +1,44 @@
+    return "{\"message\": \"Email confirmation message has been sent\"}";
+  }
+
+  @GetMapping("/activation")
```
см выше. оставляем?

IliaWizy commented on src/main/java/com/wizy/educationapp/controller/AuthorizationController.java:
```diff
@@ -0,0 +1,44 @@
+  @ResponseStatus(HttpStatus.CREATED)
+  public String registerUser(@Validated(OnCreate.class) @RequestBody UserDto dto) {
+    userService.create(userMapper.toEntity(dto));
+    return "{\"message\": \"Email confirmation message has been sent\"}";
```
обсуждали это. исправляем

IliaWizy commented on src/main/java/com/wizy/educationapp/controller/AuthorizationController.java:
```diff
@@ -0,0 +1,44 @@
+  private final VerificationTokenService verificationTokenService;
+
+  @PostMapping("/auth/sign-up")
+  @ResponseStatus(HttpStatus.CREATED)
+  public String registerUser(@Validated(OnCreate.class) @RequestBody UserDto dto) {
+    userService.create(userMapper.toEntity(dto));
+    return "{\"message\": \"Email confirmation message has been sent\"}";
+  }
+
+  @GetMapping("/activation")
+  @ResponseStatus(HttpStatus.OK)
+  public ResponseEntity<String> activateUser(@RequestParam String token) {
+    VerificationToken verificationToken = verificationTokenService.findByToken(token);
+    User user = verificationToken.getUser();
+    if (user.isActive()) {
+      return ResponseEntity.ok("User is already active.");
+    } else {
+      user.setActive(true);
+      userService.save(user);
+      return ResponseEntity.ok("User activation success.");
+    }
+  }
+}
```
исправляем по обсуждению вчера

IliaWizy commented on src/main/java/com/wizy/educationapp/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,51 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice(basePackages = "com.wizy.educationapp.controller")
+@Order(Ordered.HIGHEST_PRECEDENCE)
```
если тут глобальный, значит есть не глобальный, раз появился приоритет!? если глобальным выставить наивысший, то при равнозначном исключении у глобального будет приоритет. мы этого точно хотим? или название класса несет иной смысл?

IliaWizy commented on src/main/java/com/wizy/educationapp/dto/UserDto.java:
```diff
@@ -0,0 +1,24 @@
+
+@Data
+public class UserDto {
+  @NotNull(message = "Email must not be null", groups = OnCreate.class)
```
нот блан разве не включает параметр этой аннотации? почитай внутри нот нал и нот бланк.

IliaWizy commented on src/main/resources/application.yml:
```diff
@@ -1,6 +1,33 @@
+    driver-class-name: org.postgresql.Driver
+  jpa:
+    hibernate:
+      ddl-auto: none
```
?

IliaWizy commented on src/main/resources/liquibase/changesets/001-create-users-table-tokens-table.sql:
- на 2
- иф нот экзист

IliaWizy replied:
если что-то пойдет не так - нужна инструкция отката
после  всех манипуляций/ действий
--rollback DROP TABLE users;

поэтому в тч нужно отдельно разносить + представь, что нужно соблюдать обратную совместимость между релизами и в новой версии только одна таблица некорректно мигрирует и ее изменения не существенны, чтобы откатывать весь релиз. писать новую миграцию по всему релизу?

если нужно совместить несколько операций, например таблица + индекс + секвенс, то на каждую операцию делается отдельный ченджсет, с инкриментом (порядок выполнения)

пример:
```
--liquibase formatted sql
--add users table

--changeset username:1
CREATE TABLE IF NOT EXISTS users
(
...
);

--changeset гыуктфьу:2
CREATE INDEX indx_user_email ON users(email);

--rollback DROP INDEX indx_user_email;
--rollback DROP TABLE users;
```

IliaWizy commented on src/main/java/com/wizy/educationapp/model/VerificationToken.java:
```diff
@@ -0,0 +1,36 @@
+  @Column(name = "expiry_date")
+  private Timestamp expiryDate;
+
+  @OneToOne(cascade = {
```
вопрос был открыт. оставляешь?

IliaWizy commented on src/main/java/com/wizy/educationapp/model/VerificationToken.java:
```diff
@@ -0,0 +1,36 @@
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private int id;
```
лонг

IliaWizy commented on src/main/java/com/wizy/educationapp/model/VerificationToken.java:
```diff
@@ -0,0 +1,36 @@
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private int id;
+
+  private String token;
```
добавь uuid. в миграцию соответственно тоже, если потребуется. если на текущем этапе это долго и есть вопросы - запиши этот пункт в таску.

DMCHMB replied:
завел таску #14

IliaWizy commented on src/main/java/com/wizy/educationapp/model/User.java:
```diff
@@ -0,0 +1,22 @@
+  private String password;
+
+  @Column(columnDefinition = "BOOLEAN DEFAULT false")
+  private boolean isTutor;
+  @Column(columnDefinition = "BOOLEAN DEFAULT false")
+  private boolean isActive;
```
+ правим поле

IliaWizy commented on src/main/java/com/wizy/educationapp/model/User.java:
```diff
@@ -0,0 +1,22 @@
+
+
+@Entity
+@Data
```
нет

DMCHMB replied:
прошу помощи, как правильно сделать. Не уверен, что в новом коммите правильно реализовал

IliaWizy replied:
убрать дату и оставить гет. сет. что нужно. в чат скидывал статью про дату.

IliaWizy commented on src/main/resources/liquibase/changesets/001-create-users-table-tokens-table.sql:
```diff
@@ -0,0 +1,20 @@
+--liquibase formatted sql
+--changeset yourname:create-users-table-tokens-table
+CREATE TABLE users
+(
+    id        SERIAL PRIMARY KEY,
+    email     VARCHAR(255) UNIQUE NOT NULL,
+    full_name VARCHAR(255),
+    password  VARCHAR(255)        NOT NULL,
+    is_tutor  BOOLEAN,
+    is_active  BOOLEAN
+);
+
+CREATE TABLE verification_tokens
+(
+    id          SERIAL PRIMARY KEY,
+    user_id     INT,
+    token       VARCHAR(255),
+    expiry_date TIMESTAMP,
+    FOREIGN KEY (user_id) REFERENCES users (id)
```
ограничения точно все указаны? на код бэка не рассчитываем - бд живет своей жизнью

DMCHMB replied:
ограничения надо обсудить, задачу решенной не помечаю

IliaWizy commented on src/main/java/com/wizy/educationapp/dto/UserDto.java:
```diff
@@ -0,0 +1,24 @@
+  @NotNull(message = "Email must not be null", groups = OnCreate.class)
+  @NotBlank(message = "Email must not be blank", groups = OnCreate.class)
+  @Length(max = 255, message = "Name must be less than 255 symbols", groups = OnCreate.class)
+  private String email;
```
package jakarta.validation.constraints;
изучи пакет. по возможности - убери валидацию хайбера и замени на джакарту.

IliaWizy commented on src/main/java/com/wizy/educationapp/dto/mappers/UserMapper.java:
дима давал совет. давай прислушаемся к нему.

IliaWizy commented on src/main/java/com/wizy/educationapp/service/impl/EmailServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+      SimpleMailMessage message = new SimpleMailMessage();
+      message.setTo(user.getEmail());
+      message.setSubject("email confirmation");
+      message.setText("http://localhost:8080/api/v1/activation?token=" + token);
```
если завтра порт изменится или есть что- будем править код?

IliaWizy commented on src/main/java/com/wizy/educationapp/service/impl/EmailServiceImpl.java:
так понял, что не дописан. когда досан будет - подумаем - а если сервер который отправляет - будет висеть - мы будем его ждать? как сделать, чтобы не ждать? можем сразу это учесть?

DMCHMB replied:
записал в таску

IliaWizy commented on src/main/java/com/wizy/educationapp/service/VerificationTokenService.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface VerificationTokenService {
+
+  VerificationToken findByToken(String token);
```
отдаем токен и его же возвращаем?

DMCHMB replied:
Возвращает объект, в котором хранятся одновременно токен (строка) и user (объект). Не уверен, что верно реализовал

IliaWizy commented on src/main/java/com/wizy/educationapp/service/UserService.java:
```diff
@@ -0,0 +1,13 @@
+
+import com.wizy.educationapp.model.User;
+
+public interface UserService {
```
пора уже как-то утвердить - какие методы мы используем. предложи варианты из общеупотребляемых, чтобы на всем проекте прослеживалась одна история методов

DMCHMB replied:
Нужно определиться, не закрываю

IliaWizy commented on build.gradle.kts:
mavenLocal() первым

IliaWizy commented on src/main/java/com/wizy/educationapp/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+  @Override
+  public User getById(Long id) {
+    if (userRepository.findById(id).isEmpty()) {
+      throw new ResourceNotFoundException("User doesn't exists");
+    }
+    return userRepository.findById(id).get();
```
плохая практика. стрим апи изучаем!!! или бозово пишем, чтобы дважды не дергать репозиторий.

как вариант:
`return userRepository
            .findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("User doesn't exists"));`

IliaWizy replied:
остальные посмотри с учетом всех пометок и что вчера обсуждали.

IliaWizy reviewed (changes requested):
внеси правки

IliaWizy commented on src/main/resources/liquibase/changesets/001-create-users-table-tokens-table.sql:
```diff
@@ -0,0 +1,20 @@
+--liquibase formatted sql
+--changeset yourname:create-users-table-tokens-table
```
yourname == имя авто в таблице ченджлога

DMCHMB commented:
#12: Code refactoring:

Added:
ActivationController;
logging capability

Replaced by: 
AuthController -> RegistrationController
VerificationToken -> EmailVerificationToken
VerificationTokenService -> EmailVerificationTokenService
VerificationRepository -> EmailVerificationRepository

Added migrations to the database, added cascading token deletion on user deletion

Implemented layered architecture approach 
(controller->service->repository->service->controller)

Updated readme.md, described actions required to send email to user
